src/basic/parse-util.c

```diff
@@ -359,20 +359,34 @@ int safe_atou_full(const char *s, unsigned base, unsigned *ret_u) {
         unsigned long l;
 
         assert(s);
-        assert(base <= 16);
+        assert(SAFE_ATO_MASK_FLAGS(base) <= 16);
 
-        /* strtoul() is happy to parse negative values, and silently
-         * converts them to unsigned values without generating an
-         * error. We want a clean error, hence let's look for the "-"
-         * prefix on our own, and generate an error. But let's do so
-         * only after strtoul() validated that the string is clean
-         * otherwise, so that we return EINVAL preferably over
-         * ERANGE. */
+        /* strtoul() is happy to parse negative values, and silently converts them to unsigned values without
+         * generating an error. We want a clean error, hence let's look for the "-" prefix on our own, and
+         * generate an error. But let's do so only after strtoul() validated that the string is clean
+         * otherwise, so that we return EINVAL preferably over ERANGE. */
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_WHITESPACE) &&
+            strchr(WHITESPACE, s[0]))
+                return -EINVAL;
 
         s += strspn(s, WHITESPACE);
 
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_PLUS_MINUS) &&
+            IN_SET(s[0], '+', '-'))
+                return -EINVAL; /* Not that we check the "-" prefix again a second time below, but return a
+                                 * different error. i.e. if the SAFE_ATO_REFUSE_PLUS_MINUS we blanket refuse
+                                 * +/- prefixed integers, while if it is missing we'll just return ERANGE,
+                                 * because the string actually parses correctly, but doesn't fit in the
+                                 * return type. */
+
+        if (FLAGS_SET(base, SAFE_ATO_REFUSE_LEADING_ZERO) &&
+            s[0] == '0' && s[1] != 0)
```

Comment:
This will let '001' through.

Reply:
ouch

Reply:
hmm, actually it does not. If the `SAFE_ATO_REFUSE_LEADING_ZERO` flag is set, then for `001` the `s[0] == '0'` test will hold. but the second test `s[1] != 0` will hold too, and EINVAL is returned.

Maybe the confusion is from the fact that `0` and `'0'` look so similar?

I'll reword this as:

`s[0] && !streq(s, "0")`

That should make things clearer, no?

Reply:
`s[0] && !streq(s, "0")` is true for any non-empty string